great_ape_safe/ape_api/ebtc.py

```diff
@@ -866,3 +872,248 @@ def authority_set_authority(self, new_authority):
+        @param cdp_id The CdpId on which this operation is operated.
+        @param coll_amount The total stETH collateral amount withdrawn (reduced) for the specified Cdp.
+        """
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        # verify: cdp holds enough collateral to be withdrawn
+        (
+            _,
+            cdp_id_coll,
+            _,
+            _,
+            _,
+            _,
+        ) = self.cdp_manager.Cdps(cdp_id)
+        assert cdp_id_coll > coll_amount
+
+        feed_price = self.price_feed.fetchPrice.call()
+        prev_icr = self.cdp_manager.getSyncedICR(cdp_id, feed_price)
+        prev_tcr = self.cdp_manager.getSyncedTCR(feed_price)
+        prev_coll_balance = self.cdp_manager.getCdpCollShares(cdp_id)
+
+        # 1. decreased collateral in target cdp id
+        self.borrower_operations.withdrawColl(cdp_id, coll_amount, cdp_id, cdp_id)
+
+        # 2. assertions:
+
+        # 2.1. icr and tcr had decreased
+        assert self.cdp_manager.getSyncedICR(cdp_id, feed_price) < prev_icr
+        assert self.cdp_manager.getSyncedTCR(feed_price) < prev_tcr
+
+        # 2.2 collateral in cdp at storage has decreased
+        assert self.cdp_manager.getCdpCollShares(cdp_id) < prev_coll_balance
+
+    def cdp_repay_debt(self, cdp_id, debt_repay_amount):
+        """
+        @dev Repays the received eBTC token to the specified Cdp, thus reducing its debt accounting.
+        @param cdp_id The CdpId on which this operation is operated.
+        @param debt_repay_amount The total eBTC debt amount repaid for the specified Cdp.
+        """
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        # verify: check debt caller balance
+        prev_debt_balance = self._assert_debt_balance(debt_repay_amount)
+
+        feed_price = self.price_feed.fetchPrice.call()
+        prev_icr = self.cdp_manager.getSyncedICR(cdp_id, feed_price)
+
+        # 1. repay debt
+        self.borrower_operations.repayDebt(cdp_id, debt_repay_amount, cdp_id, cdp_id)
+
+        # 2. assertions
+
+        # 2.1. debt balance decreased
+        assert self.ebtc_token.balanceOf(self.safe.address) < prev_debt_balance
+
+        # 2.2. icr should improved
+        assert self.cdp_manager.getSyncedICR(cdp_id, feed_price) > prev_icr
+
+    def cdp_withdraw_debt(self, cdp_id, debt_withdrawable_amount):
+        """
+        @dev Withdraws the amount of eBTC token from the specified Cdp, thus increasing its debt accounting.
+        @param cdp_id The CdpId on which this operation is operated.
+        @param debt_withdrawable_amount The total debt collateral amount increased for the specified Cdp.
+        """
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        # verify: existing debt in cdp id is greater than amount to wd
+        debt_before = self.cdp_manager.getCdpDebt(cdp_id)
```

Comment:
I think some off-chain ICR reduction checks would make sense here to make sure that we don't push lower unexpected thresholds. My suggestion:
- I would assert that the new ICR is not below Liquidation TH (110%) to avoid the revert
- I would assert that we are not in RM (it is not allowed to only withdraw debt while in RM.
- I would perhaps output the expected new ICR and require console confirmation for proceeding (could catch unexpected values if price swings ocurred from the moment the decision was made to the moment it was posted as well as typos on amounts).

What do you think?

Reply:
Now that I think about it, perhaps it makes sense to have the ICR change confirmation for all adjustment operations. Just to be safe (both reductions and increasements).

Reply:
aimed to encapsulate feedback in this commit 9e843c88de21d5a134f48a8440a9a8386f6890a2

agree on the asserts and logging on ops specially worsening up the icr for the target cpd id

Reply:
LGTM